### PR TITLE
feat(utils): 增加微信公众号文章音频部分处理逻辑，修正公众号文章中引入的视频内容

### PR DESCRIPTION
**Pull request by rpsh**
## 该 PR 相关 Issue / Involved issue
无相关 issue
Close #

## 完整路由地址 / Example for the proposed route(s)
```route
/wechat/feeddd/6131e1401269c358aa0e1867
/wechat/feeddd/615b3b7a1269c358aa15a17b
/wechat/feeddd/625d00c2a4ca6e10e36bb689
```

## 新 RSS 检查列表 / New RSS Script Checklist
  
- [ ] 新的路由 New Route
  - [ ] 跟随 [v2 路由规范](https://docs.rsshub.app/joinus/script-standard.html) Follows [v2 Script Standard](https://docs.rsshub.app/en/joinus/script-standard.html)
- [ ] 文档说明 Documentation
  - [ ] 中文文档 CN
  - [ ] 英文文档 EN
- [ ] 全文获取 fulltext
  - [ ] 使用缓存 Use Cache
- [ ] 反爬/频率限制 anti-bot or rate limit?
  - [ ] 如果有, 是否有对应的措施? If yes, do your code reflect this sign?
- [ ] [日期和时间](https://docs.rsshub.app/joinus/pub-date.html) [date and time](https://docs.rsshub.app/en/joinus/pub-date.html)
  - [ ] 可以解析 Parsed
  - [ ] 时区调整 Correct TimeZone
- [ ] 添加了新的包 New package added
- [ ] `Puppeteer`

## 说明 / Note
- 增加微信公众号文章中音频部分处理逻辑
- 修正公众号文章中引入的视频 <iframe> 内容

### CODE REVIEW

File: lib/utils/wechat-mp.js [hunk 1]
```diff
@@ -178,24 +312,28 @@ const fetchArticle = async (ctx, url, bypassHostCheck = false) => {
...
+        const rootHtml = response.data;
+        const $ = cheerio.load(rootHtml);
 
         const title = ($('meta[property="og:title"]').attr('content') || '').replace(/\\r/g, '').replace(/\\n/g, ' ');
         const author = $('meta[name=author]').attr('content');
         let summary = $('meta[name=description]').attr('content');
         summary = summary !== title ? summary : '';
-        let description = fixArticleContent($('#js_content'));
+        const headImage = getAccountHeadImage(rootHtml);
+        const voiceInfo = detectVoiceInfo(rootHtml);
```
Comment (Rongronggg9) [inline lib/utils/wechat-mp.js:323]: Please do not aggressively match the whole HTML, but firstly using cheerio to select that part may contain these resources before performing matching.

File: lib/utils/wechat-mp.js [hunk 2]
```diff
@@ -178,24 +312,28 @@ const fetchArticle = async (ctx, url, bypassHostCheck = false) => {
...
+        let description = await fixArticleContent($('#js_content'), false, voiceInfo);
 
         // No article get or article is too short, try the original url
         const originalUrl = detectOriginalArticleUrl($);
         if (originalUrl) {
             // try to fetch the description from the original article
             const originalResponse = await got(normalizeUrl(originalUrl, bypassHostCheck));
             const original$ = cheerio.load(originalResponse.data);
-            description += fixArticleContent(original$('#js_content'));
+            const originalVoiceInfo = detectVoiceInfo(originalResponse.data);
```
Comment (Rongronggg9) [inline lib/utils/wechat-mp.js:332]: Please do not aggressively match the whole HTML, but firstly using cheerio to select that part may contain these resources before performing matching.

File: lib/utils/wechat-mp.js [hunk 3]
```diff
@@ -103,14 +185,66 @@ const fixArticleContent = (html, skipImg = false) => {
...
+    });
+
+    // fix single voice article
+    // example: https://mp.weixin.qq.com/s/FY6yQC_e4NMAxK0FBr6jwQ
+    $('#voice_parent').each((_, voice) => {
+        voice = $(voice);
+        if (voiceInfo) {
+            voice.html(renderAudio(voiceInfo));
+        }
+    });
```
Comment (Rongronggg9) [inline lib/utils/wechat-mp.js:238]:
```suggestion
    // fix single voice article
    // example: https://mp.weixin.qq.com/s/FY6yQC_e4NMAxK0FBr6jwQ
    if (voiceInfo) {
    	$('#voice_parent').each((_, voice) => {
        	voice = $(voice);
            voice.html(renderAudio(voiceInfo));
    	});
    }
```

File: lib/utils/wechat-mp.js [hunk 4]
```diff
@@ -178,24 +312,28 @@ const fetchArticle = async (ctx, url, bypassHostCheck = false) => {
...
             // try to fetch the description from the original article
             const originalResponse = await got(normalizeUrl(originalUrl, bypassHostCheck));
             const original$ = cheerio.load(originalResponse.data);
-            description += fixArticleContent(original$('#js_content'));
+            const originalVoiceInfo = detectVoiceInfo(originalResponse.data);
+            description += await fixArticleContent(original$('#js_content'), false, originalVoiceInfo);
         }
 
-        const sourceUrl = detectSourceUrl($);
+        const sourceUrl = detectSourceUrl(rootHtml);
```
Comment (Rongronggg9) [inline lib/utils/wechat-mp.js:336]: Please do not aggressively match the whole HTML, but firstly using cheerio to select that part may contain these resources before performing matching.

File: lib/utils/wechat-mp.js [hunk 5]
```diff
@@ -103,14 +185,66 @@ const fixArticleContent = (html, skipImg = false) => {
...
+            voice.html(renderAudio(voiceInfo));
+        }
+    });
+
+    // fix single picture/vocice article
     // example: https://mp.weixin.qq.com/s/4p5YmYuASiQSYFiy7KqydQ
     $('script').each((_, script) => {
         script = $(script);
-        const matchs = script.html().match(/document\.getElementById\('js_image_desc'\)\.innerHTML = "(.*)"\.replace/);
+        const matchs = script.html().match(/document\.getElementById\('(?:js_image_desc|js_audio_desc)'\)\.innerHTML = (?:'|")(.*)(?:'|")\.replace/);
```
Comment (Rongronggg9) [inline lib/utils/wechat-mp.js:244]:
```suggestion
        const matchs = script.html().match(/document\.getElementById\('(?:js_image_desc|js_audio_desc)'\)\.innerHTML = ['"](.*)['"]\.replace/);
```

File: lib/utils/wechat-mp.js [hunk 6]
```diff
@@ -247,6 +386,10 @@ const finishArticleItem = async (ctx, item, setMpNameAsAuthor = false, skipLink
     if (!skipLink) {
         item.link = link || item.link;
     }
+    if (voiceInfo && voiceInfo.voiceId) {
+        item.enclosure_url = `https://res.wx.qq.com/voice/getvoice?mediaid=${voiceInfo.voiceId}`;
```
Comment (Rongronggg9) [inline lib/utils/wechat-mp.js:390]: Please do not write a URL twice.

File: lib/utils/wechat-mp.js [hunk 7]
```diff
@@ -49,14 +55,89 @@ const detectOriginalArticleUrl = ($) => {
...
+ *
+ * @param {object} options - The info of voice.
+ * @param {string} options.voiceId - The voice id of voice.
+ * @param {string} [options.name] - The name of voice.
+ * @param {string} [options.duration] - The duration of voice .
+ * @return {string} - The <audio> html, a string.
+ */
+const renderAudio = (options = {}) => {
+    if (options.voiceId) {
+        return `<audio controls="controls" preload="auto" duration="${options.duration || ''}"  style="width:100%" src="https://res.wx.qq.com/voice/getvoice?mediaid=${options.voiceId}" title="${options.name || ''}"></audio>`;
```
Comment (Rongronggg9) [inline lib/utils/wechat-mp.js:106]: Please do not write a URL twice.

**Review by Rongronggg9 — CHANGES_REQUESTED** [review]

File: lib/utils/wechat-mp.js [hunk 8]
```diff
@@ -49,14 +55,89 @@ const detectOriginalArticleUrl = ($) => {
...
+ * @param {string} html - The html text of the article.
+ * @return {(string|null)} - A head image url.
+ */
+const getAccountHeadImage = (html = '') => {
+    const matchs = html.match(/hd_head_img = "([^"]+)"/);
     if (matchs) {
         return matchs[1];
     }
+
+    const matchs2 = html.match(/getXmlValue\('hd_head_img.DATA'\) \|\| '' : '([^']+)'/);
```
Comment (Rongronggg9) [inline lib/utils/wechat-mp.js:123]:
```suggestion
    const matchs2 = html.match(/getXmlValue\('hd_head_img.DATA'\) *\|\| *'' *: *'([^']+)'/);
```
To make it more adaptive.

**Review by Rongronggg9 — CHANGES_REQUESTED** [review]

**Comment by DIYgod** [comment]
pr 时间过长，将先被关闭，如果还有疑问可以随时在下面继续讨论

**Pull request closed without merging**